xds/src/main/java/io/grpc/xds/internal/sts/StsCredentials.java

```diff
@@ -0,0 +1,164 @@
+  private transient HttpTransportFactory transportFactory;
+
+  private StsCredentials(
+      String identityTokenEndpoint,
+      String audience,
+      String sourceCredentialsFileLocation,
+      HttpTransportFactory transportFactory) {
+    this.identityTokenEndpoint = identityTokenEndpoint;
+    this.audience = audience;
+    this.sourceCredentialsFileLocation = sourceCredentialsFileLocation;
+    this.transportFactory = transportFactory;
+  }
+
+  /**
+   * Creates an StsCredentials.
+   *
+   * @param identityTokenEndpoint  URL of the token exchange service to use.
+   * @param audience Audience to use in the STS request.
+   * @param sourceCredentialsFileLocation file-system location that contains the
+   *                                      source creds e.g. JWT contents.
+   */
+  public static StsCredentials create(
+      String identityTokenEndpoint, String audience, String sourceCredentialsFileLocation) {
+    return create(
+        identityTokenEndpoint,
+        audience,
+        sourceCredentialsFileLocation,
+        getFromServiceLoader(HttpTransportFactory.class, defaultHttpTransportFactory));
+  }
+
+  @VisibleForTesting
+  static StsCredentials create(
+      String identityTokenEndpoint,
+      String audience,
+      String sourceCredentialsFileLocation,
+      HttpTransportFactory transportFactory) {
+    return new StsCredentials(
+        identityTokenEndpoint, audience, sourceCredentialsFileLocation, transportFactory);
+  }
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    AccessToken tok = getSourceAccessTokenFromFileLocation();
+
+    HttpTransport httpTransport = this.transportFactory.create();
+    JsonObjectParser parser = new JsonObjectParser(JacksonFactory.getDefaultInstance());
+
+    HttpRequestFactory requestFactory = httpTransport.createRequestFactory();
+    GenericUrl url = new GenericUrl(identityTokenEndpoint);
+
+    Map<String, String> params = new HashMap<>();
+    params.put("grant_type", "urn:ietf:params:oauth:grant-type:token-exchange");
+    params.put("subject_token_type", "urn:ietf:params:oauth:token-type:jwt");
+    params.put("requested_token_type", "urn:ietf:params:oauth:token-type:access_token");
+    params.put("subject_token", tok.getTokenValue());
+    params.put("scope", CLOUD_PLATFORM_SCOPE);
+    params.put("audience", audience);
+    HttpContent content = new JsonHttpContent(parser.getJsonFactory(), params);
+    HttpRequest request = requestFactory.buildPostRequest(url, content);
+    request.setParser(parser);
+
+    HttpResponse response = null;
+    try {
+      response = request.execute();
+    } catch (IOException e) {
+      throw new IOException("Error requesting access token", e);
+    }
+
+    if (response.getStatusCode() != HttpStatusCodes.STATUS_CODE_OK) {
+      throw new IOException("Error getting access token " + getStatusString(response));
+    }
+
+    GenericData responseData = response.parseAs(GenericData.class);
```

Comment:
Surround this line with `try{}` and next line with `finally{}`?

Reply:
This pattern was copied from other similar implementations in google-auth e.g. https://github.com/googleapis/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java#L264 

Many other implementations don't even disconnect! Anyway I'll use try-finally as suggested!